CONTRIBUTING.md

```diff
@@ -57,3 +57,325 @@ If the changes in your PR do not conflict with any of the existing code in the p
+    ```
+1. Go to the Consul `api` package for the config entry, e.g. https://github.com/hashicorp/consul/blob/master/api/config_entry_gateways.go
+1. Copy the top-level fields over into the `Spec` struct except for
+   `Kind`, `Name`, `Namespace`, `Meta`, `CreateIndex` and `ModifyIndex`. In this
+   example, the top-level fields remaining are `TLS` and `Listeners`:
+   
+    ```go
+    // IngressGatewaySpec defines the desired state of IngressGateway
+    type IngressGatewaySpec struct {
+        // TLS holds the TLS configuration for this gateway.
+        TLS GatewayTLSConfig
+        // Listeners declares what ports the ingress gateway should listen on, and
+        // what services to associated to those ports.
+        Listeners []IngressListener
+    }
+    ```
+1. Copy the structs over that are missing, e.g. `GatewayTLSConfig`, `IngressListener`.
+1. Set `json` tags for all fields using camelCase starting with a lowercase letter:
+    ```go
+        TLS GatewayTLSConfig `json:"tls"`
+    ```
+   Note that you should use the fields name, e.g. `tls`, not the struct name, e.g. `gatewayTLSConfig`.
+   Remove any `alias` struct tags.
+1. If the fields aren't documented, document them using the Consul docs as a reference.
+
+### Spec Methods
+1. Run `make ctrl-generate` to implement the deep copy methods.
+1. Implement all the methods for `ConfigEntryResource` in the `_types.go` file. If using goland you can
+   automatically stub out all the methods by using Code -> Generate -> IngressGateway -> ConfigEntryResource.
+1. Use existing implementations of other types to implement the methods. We have to
+   copy their code because we can't use a common struct that implements the methods
+   because that messes up the CRD code generation. 
+   
+   You should be able to follow the other "normal" types. The non-normal types
+   are `ServiceIntention` and `ProxyDefault` because they have special behaviour
+   around being global or their spec not matching up with Consul's directly.
+1. When you get to `ToConsul` and `Validate` you'll need to actually think
+   about the implementation instead of copy/pasting and doing a simple replace.
+1. For `ToConsul`, the pattern we follow is to implement `toConsul()` methods
+   on each sub-struct. You can see this pattern in the existing types.
+1. For `Validate`, we again follow the pattern of implementing the method on
+   each sub-struct. You'll need to read the Consul documentation to understand
+   what validation needs to be done.
+   
+   Things to keep in mind:
+   1. Re-use the `sliceContains` and `notInSliceMessage` helper methods where applicable.
+   1. If the invalid field is an entire struct, encode as json (look for `asJSON` for an example).
+   1. `validateNamespaces` should be a separate method.
+   1. If the field can have a `nil` pointer, check for that, e.g.
+        ```go
+        func (in *ServiceRouteHTTPMatchHeader) validate(path *field.Path) *field.Error {
+            if in == nil {
```

Comment:
should this be `in` or `path` ?

Reply:
It needs to be `in` because if the CRD is:
```
spec:
  routes:
    - match:
        pathPrefix: ....
```

Then `ServiceRouter.Routes[0].match.header` will be `nil` but we blindly call
```go
	errs = append(errs, in.Match.validate(path.Child("match"))...)
```

So we need to check inside `validate` if our pointer caller is `nil`. We could also check outside of the call to `validate()` but this is the pattern we've followed.

`path` won't be `nil` because we're always passing it properly.